<font color="blue">Привет. Давай смотреть как у тебя с линейной регрессией

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

## 1. Загрузка данных

In [5]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

In [6]:
df = pd.read_csv('/datasets/insurance.csv')

In [7]:
df.head()

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

In [8]:
df.describe()

Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
Пол                  5000 non-null int64
Возраст              5000 non-null float64
Зарплата             5000 non-null float64
Члены семьи          5000 non-null int64
Страховые выплаты    5000 non-null int64
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


In [10]:
df.shape

(5000, 5)

Пропусков нет. Типы данных не требуют преобразований.

Создаю переменные для целевого признака и остальных признаков.

In [11]:
features = df.drop('Страховые выплаты', axis=1)
target = df['Страховые выплаты']

## 2. Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

Признаки умножают на обратимую матрицу. Изменится ли качество линейной регрессии? (Её можно обучить заново.)
- a. Изменится. Приведите примеры матриц.
- b. Не изменится. Укажите, как связаны параметры линейной регрессии в исходной задаче и в преобразованной.

**Ответ:** b. Не изменится.

**Обоснование:**

$Z$ - обратимая матрица

$a'$ - предсказания после преобразования

$w'$ — вектор весов линейной регрессии после преобразования

$$
a' = XZw'
$$

$$
w' = \arg\min_w MSE(XZw, y)
$$

$$
w' = ((XZ)^T (XZ))^{-1} (XZ)^T y
$$

$$
XZw' = (XZ) ((XZ)^T(XZ))^{-1} (XZ)^T y = (XZ) (Z^T X^T X Z)^{-1} (XZ)^T y 
$$

Так как матрица Z обратимая, то

$$
XZw' = (XZ) Z^{-1} (X^T X)^{-1} Z^T (XZ)^T y = X(X^T X)^{-1} X y
$$

Соответственно умножение на обратимую матрицу не повлияет на качество модели

Как связаны параметры линейной регрессии?
$$
w' = ((XZ)^T (XZ))^{-1} (XZ)^T y = (Z^T X^T M_1 Z)^{-1} (XZ)^T y = 
$$
$$
Z^{-1} (X^T X)^{-1} Z^{-T} Z^T X^T y = Z^{-1} w
$$

<font color="blue">Численный эксперимент штука хорошая, но здесь его применять не стоит. Да и сам ответ не корректен. Поиграйся чуток формулами и получишь красивое доказательство верного решения

## 3. Алгоритм преобразования

**Алгоритм**

- $M_1$ — исходная матрица признаков размера m x n

- $Z$ — случайная обратимая матрица размера n x n

- $M_2$ — преобразованная матрица признаков

$$
M_2 = M_1 Z
$$

Обратное преобразование 
$$
M_1 = M_2 Z^{-1}
$$

**Обоснование**

- Как было описано выше, умножение матрицы на обратимую матрицу не изменяет качество линейной регрессии.
- Если матрица обратимая, то $M_2 = M_1 Z$, можно представить как $M_1 = M_2 Z^{-1}$

## 4. Проверка алгоритма

Расчет метрики R2 до преобразования.

In [66]:
model.fit(features, target)
pred = model.predict(features)
print(r2_score(target, pred))

0.42494550286668


Значение метрики R2 до преобразования 0.425

Преобразование матрицы признаков по описанному алгоритму

In [67]:
m1 = features.values

In [70]:
z = np.random.rand(m1.shape[1],m1.shape[1])

In [73]:
m2 = m1.dot(z)

0.42494550286668364


In [74]:
m2

array([[20695.94946386, 11638.09424833, 23476.91749767, 32212.3805265 ],
       [15868.10907445,  8919.18346126, 17997.06707555, 24692.79556444],
       [ 8771.84842445,  4929.47468055,  9948.06781182, 13649.13611886],
       ...,
       [14138.79108368,  7953.09459819, 16040.28004308, 22009.04448513],
       [13642.09855818,  7672.95813881, 15475.60038918, 21233.3560991 ],
       [16936.09284715,  9525.34478689, 19213.0333412 , 26362.20338155]])

Расчет метрики R2 после преобразования

In [78]:
model.fit(m2, target)
pred = model.predict(m2)
print(r2_score(target, pred))

0.42494550286668364


Обратное преобразование

In [76]:
m1 = m2.dot(np.linalg.inv(z))
m1

array([[ 1.00000000e+00,  4.10000000e+01,  4.96000000e+04,
         1.00000000e+00],
       [-3.22553090e-12,  4.60000000e+01,  3.80000000e+04,
         1.00000000e+00],
       [ 1.27225744e-11,  2.90000000e+01,  2.10000000e+04,
        -4.74295246e-12],
       ...,
       [-1.13964799e-11,  2.00000000e+01,  3.39000000e+04,
         2.00000000e+00],
       [ 1.00000000e+00,  2.20000000e+01,  3.27000000e+04,
         3.00000000e+00],
       [ 1.00000000e+00,  2.80000000e+01,  4.06000000e+04,
         1.00000000e+00]])

После преобразования значение метрики R2 не изменилось. Качество модели не ухудшилось. Обратное преобразование позволяет восстановить исходные данные.

<font color="blue">Тут замечание такое - ты опираешься на всю выборку. А что если она изменится твой алгоритм будет работать? Подумай как тут можно сделать универсальное решение? Ну и сам алгоритм масштабирования - такое.  Давай подумай как можно сделать так чтобы сами фичи были как бы спутанны между собой.  Более того, в нотации есть $M_1+1$ это как? ты скаляр складываешь с матрицей.

## Чек-лист проверки

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные загружены
- [x]  Выполнен шаг 2: получен ответ на вопрос об умножении матриц
    - [x]  Указан правильный вариант ответа
    - [x]  Вариант обоснован
- [x]  Выполнен шаг 3: предложен алгоритм преобразования
    - [x]  Алгоритм описан
    - [x]  Алгоритм обоснован
- [x]  Выполнен шаг 4: алгоритм проверен
    - [x]  Алгоритм реализован
    - [x]  Проведено сравнение качества моделей до и после преобразования